File: examples/Perl/rrbroker.pl
# Simple request-reply broker in Perl
# Uses AnyEvent to poll the sockets

use strict;
use warnings;
use v5.10;

use ZMQ::FFI;
use ZMQ::FFI::Constants qw(ZMQ_ROUTER ZMQ_DEALER);

use AnyEvent;
use EV;

# Prepare our context and sockets
my $context  = ZMQ::FFI->new();
my $frontend = $context->socket(ZMQ_ROUTER);
my $backend  = $context->socket(ZMQ_DEALER);

$frontend->bind('tcp://*:5559');
$backend->bind('tcp://*:5560');

# Switch messages between sockets

my $frontend_poller = AE::io $frontend->get_fd, 0, sub {
    while ($frontend->has_pollin) {
        my @message = $frontend->recv_multipart();
        $backend->send_multipart(\@message);
    }
};

my $backend_poller = AE::io $backend->get_fd, 0, sub {
    while ($backend->has_pollin) {
        my @message = $backend->recv_multipart();
        $frontend->send_multipart(\@message);
    }
};

EV::run;

File: examples/Python/asyncio_ioloop/rrbroker.py
#!/usr/bin/env python

"""
synopsis:
    Simple request-reply broker.
    Author: Lev Givon <lev(at)columbia(dot)edu>
    Modified for async/ioloop: Dave Kuhlman <dkuhlman(at)davekuhlman(dot)org>
usage:
    python rrbroker.py
notes:
    To run this, start rrbroker.py, any number of instances of rrworker.py,
    and rrclient.py.
"""

import sys
import zmq
from zmq.asyncio import Context, Poller, ZMQEventLoop
import asyncio


@asyncio.coroutine
def run_broker(context):
    # Prepare our context and sockets
    frontend = context.socket(zmq.ROUTER)
    backend = context.socket(zmq.DEALER)
    frontend.bind("tcp://*:5559")
    backend.bind("tcp://*:5560")
    # Initialize poll set
    poller = Poller()
    poller.register(frontend, zmq.POLLIN)
    poller.register(backend, zmq.POLLIN)
    # Switch messages between sockets
    while True:
        socks = yield from poller.poll()
        socks = dict(socks)
        if socks.get(frontend) == zmq.POLLIN:
            message = yield from frontend.recv_multipart()
            print('received from frontend: {}'.format(message))
            yield from backend.send_multipart(message)
        if socks.get(backend) == zmq.POLLIN:
            message = yield from backend.recv_multipart()
            print('received from backend: {}'.format(message))
            yield from frontend.send_multipart(message)


@asyncio.coroutine
def run(loop):
    context = Context()
    yield from run_broker(context)


def main():
    args = sys.argv[1:]
    if len(args) != 0:
        sys.exit(__doc__)
    try:
        loop = ZMQEventLoop()
        asyncio.set_event_loop(loop)
        loop.run_until_complete(run(loop))
    except KeyboardInterrupt:
        print('\nFinished (interrupted)')


if __name__ == '__main__':
    main()
